### ISSUE TITLE
Advice needed 

### ISSUE THREAD

**Issue by nOmi7** [processone/ejabberd#1521]
> What version of ejabberd are you using?
17.01.45

> What operating system (version) are you using?
deb

> How did you install ejabberd (source, package, distribution)?
source

I need experts view on the following things

1) We are using mysql to store the roster , privacy , lastseen, and userinfo for auth ...
     Whenever session is created i am fetching the info from c2s state map for pres_f , pres_t and privacy list to avoid freq db calls. but whenever the user is offline I am hitting db for each call that is happening from peers , to avoid this i created ets table and caching it there . so until server is not restarted  i am good .. the question i have is it good thing to do  have less load now so it works without any issues , what effect will have when sessions increases ? does ETS have any limit on storing records ? when we do clustering will it have any effect ?

2) If we drop mysql and go on mnesia how do we store the data when mnesia grows beyond 2 gig limit it has ? how does mneisa behave in terms of clustering do i have to make some specific settings ? 

We have never done cluster but will be doing soon.

Any other approach that can make process fast is welcomed.
Sorry for posting it under issues however I did not know any place for this kinda questions.

Thanks

**Comment by zinid** [comment]
1) ETS doesn't have any limits, however, like with any cache, you should keep it's memory under control. Probably `cache_tab` will fit better in your case.
3) Usually cache should not be shared across nodes, so you're fine here.
2) Mnesia just sucks, don't even consider it.

**Comment by nOmi7** [comment]
Well @zinid _"Mnesia just sucks, don't even consider it."_  thats gonna give me an heart attack man :) 
all my offline store is on mnesia for faster message storage and retrieval  but i already faced the problem of 2gig limit so we started controlling data ... not sure how does whatsapp manage this so smoothly they said they use mnesia everywhere .

for avoiding messages lost all the message that are coming before forwarding i am storing it in mnesia table once i get the ack of delivery i  am deleting the rec ... I have to replace mnesia is there any other way i could do this ??

does cache_tab provide mechanism to write /dump to persistent storage on demand or before we shut down the node ?

**Comment by zinid** [comment]
> not sure how does whatsapp manage this so smoothly they said they use mnesia everywhere

I have no idea how it's possible to use Mnesia under such high load. We (ProcessOne) wrote our own database (p1db) for this: it has built-in caching and replication support.

> for avoiding messages lost all the message that are coming before forwarding i am storing it in mnesia table once i get the ack of delivery i am deleting the rec ... I have to replace mnesia is there any other way i could do this ??

No. If you don't need to share this Mnesia table accross nodes, you're fine (I assume 2gb limit is not a problem in this case).

> does cache_tab provide mechanism to write /dump to persistent storage on demand or before we shut down the node ?

No, writes are always performed, only reads get cached.

Let me rephrase: Mnesia has the following problem:
1) It performs very badly in clustering environment
2) It has 2gb limit problem
3) It is hard to maintain (hard to change node name for instance).

If you don't need clustering or don't need to store anything on disc, then Mnesia is pretty fine.

**Comment by nOmi7** [comment]
looks like its using treaps underneath and not ets ? ,  is cache_tab shared across the nodes in cluster or can it be shared ? I need one shared in-mem db across ejabberd cluster

**Comment by zinid** [comment]
> looks like its using treaps underneath and not ets ?

Yes, in order to efficiently delete staled records.

> is cache_tab shared across the nodes in cluster or can it be shared ?

Yes, via `dirty_dist_insert()` and `dirty_dist_delete()`.

**Comment by nOmi7** [comment]
super ... thanks ..

**Comment by nOmi7** [comment]
in case of dirty_dist_insert() where / how should i be creating a table ? if i add it in mod_xxx init() all nodes on start will creates it table ? what is the right way to create dist table

**Comment by zinid** [comment]
You should create the caching table first using `cache_tab:new()` on every node (just put this in `init()` of your `mod_xxx` and all nodes on start will create and share it). Then you can call `dirty_dist_insert()`.
